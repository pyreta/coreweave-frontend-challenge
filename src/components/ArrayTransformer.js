import React, { useState, useEffect } from "react";
import List from "./List";
import Button from "./Button";
import { shuffle, swapCase, makeRandomString } from "../utils/transformations";

export const arrayChangingFunctions = {
  "Scramble each element": { fn: shuffle, method: "map" },
  "Swap case of each element": { fn: swapCase, method: "map" },
  "Add random character to each element": {
    fn: (x) => `${x}${makeRandomString(1)}`,
    method: "map",
  },
  "Remove elements with a capital B": {
    fn: (x) => !x.includes("B"),
    method: "filter",
  },
};

const ArrayTransformer = ({inputArray, outputArray, setOutputArray, setRuntime}) => {
  const [changes, setChanges] = useState([]);
  const [loading, setLoading] = useState(false);

  const firstEl = outputArray[0];
  const len = outputArray.length;

  useEffect(() => {
    setLoading(false);
  }, [firstEl, len]);

  useEffect(() => {
    if (loading && inputArray) {
      let startTime = performance.now();
      let newArr = inputArray;
      changes.forEach((change, i) => {
        const { method, fn } = arrayChangingFunctions[change];
        newArr = newArr[method](fn);
      });
      let finishTime = performance.now();
      setRuntime(finishTime - startTime);
      setOutputArray(newArr);
    }
  }, [loading, changes, inputArray, setOutputArray, setRuntime]);


  return (
    <>
      <div>Transformations</div>
      <div>
        {Object.keys(arrayChangingFunctions).map((text) => (
          <button key={text} text={text} onClick={() => setChanges([...changes, text])}>{text}</button>
        ))}
      </div>
      <List listItems={changes} />
      <Button
        text={loading ? "Applying your changes..." : "Apply changes to array"}
        disabled={loading}
        onClick={() => {
          if (inputArray.length > 0 && changes.length > 0) setLoading(true);
        }}
      />
    </>
  )
};

export default ArrayTransformer;

