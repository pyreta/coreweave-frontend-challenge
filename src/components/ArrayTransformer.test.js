import { render, fireEvent } from '@testing-library/react';
import ArrayTransformer from './ArrayTransformer';

describe("ArrayTransformer", () => {
  const setOutputArray = jest.fn();
  const setRuntime = jest.fn();

  it('matches snapshot', () => {
    const container = render(<ArrayTransformer inputArray={[]} outputArray={[]} setOutputArray={setOutputArray} setRuntime={setRuntime} />).container;
    expect(container).toMatchSnapshot();
  });

  it('renders the correct number of buttons', () => {
    const { getAllByRole } = render(<ArrayTransformer inputArray={[]} outputArray={[]} setOutputArray={setOutputArray} setRuntime={setRuntime} />);
    const buttons = getAllByRole('button');
    expect(buttons).toHaveLength(5);
  })

  it('adds to changes list when button is clicked', () => {
    const { getAllByRole } = render(<ArrayTransformer inputArray={["1","2","3"]} outputArray={[]} setOutputArray={setOutputArray} setRuntime={setRuntime} />);
    const transformButtons = getAllByRole('button').slice(0,4);
    
    transformButtons.forEach((button) => fireEvent.click(button));

    let listItems = getAllByRole('listitem');
    expect(listItems).toHaveLength(4);
  });

  it('renders correct button text when loading transform', () => {
    const { getAllByRole, getByRole } = render(<ArrayTransformer inputArray={["1","2","3"]} outputArray={[]} setOutputArray={setOutputArray} setRuntime={setRuntime} />);
    const transformButtons = getAllByRole('button').slice(0,4);

    transformButtons.forEach((button) => {
      fireEvent.click(button);
      fireEvent.click(button);
      fireEvent.click(button);
    });

    const applyChangesButton = getByRole('button', {
      name: /Apply changes to array/i
    })
    fireEvent.click(applyChangesButton);

    expect(applyChangesButton).toHaveTextContent('Applying your changes...');
  });
});
