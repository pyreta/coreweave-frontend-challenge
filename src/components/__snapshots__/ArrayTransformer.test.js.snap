// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrayTransformer matches snapshot 1`] = `
<div>
  <div>
    Transformations
  </div>
  <div>
    <button
      text="Scramble each element"
    >
      Scramble each element
    </button>
    <button
      text="Swap case of each element"
    >
      Swap case of each element
    </button>
    <button
      text="Add random character to each element"
    >
      Add random character to each element
    </button>
    <button
      text="Remove elements with a capital B"
    >
      Remove elements with a capital B
    </button>
  </div>
  <ol>
    <div>
      No filters selected
    </div>
  </ol>
  <button>
    Apply changes to array
  </button>
</div>
`;
